### ISSUE TITLE
[jvm-packages] xgboost4j-spark test fails with "AssertError:maxdf must be smaller than FDSETSIZE"

### ISSUE THREAD

**Issue by gtsoukas** [dmlc/xgboost#2736]
## Environment info
Operating System:

SLES11 SP4

```
uname -a
Linux xxxxxxxx 3.0.101-84-default #1 SMP Tue Oct 18 10:32:51 UTC 2016 (15251d6) x86_64 x86_64 x86_64 GNU/Linux

lsb_release -a
LSB Version:    core-2.0-noarch:core-3.2-noarch:core-4.0-noarch:core-2.0-x86_64:core-3.2-x86_64:core-4.0-x86_64:desktop-4.0-amd64:desktop-4.0-noarch:graphics-2.0-amd64:graphics-2.0-noarch:graphics-3.2-amd64:graphics-3.2-noarch:graphics-4.0-amd64:graphics-4.0-noarch
Distributor ID: SUSE LINUX
Description:    SUSE Linux Enterprise Server 11 (x86_64)
Release:        11
Codename:       n/a

```

Compiler:

```
java -version
java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

-- The C compiler identification is GNU 4.3.4
-- The CXX compiler identification is GNU 5.3.1

cmake -version
cmake version 3.9.2
```

`xgboost`  commit hash:
git rev-parse HEAD

c09204fa70c665ad8c1e265b391847e8bc6e1dc0

Pleae find full log attached.
[2017-09-21_xgboost_jvm_build_failure.txt](https://github.com/dmlc/xgboost/files/1320558/2017-09-21_xgboost_jvm_build_failure.txt)

## Steps to reproduce

```
git clone --recursive https://github.com/dmlc/xgboost

cd xgboost
git rev-parse HEAD

c09204fa70c665ad8c1e265b391847e8bc6e1dc0

cd jvm-pakcages
mvn package
```

## What have you tried?

```
git checkout v0.60
mvn clean
mvn package
```

**Comment by superbobry** [comment]
Can you provide the full build log?

**Comment by gtsoukas** [comment]
Isn't that the file attached above *2017-09-21_xgboost_jvm_build_failure.txt*?
https://github.com/dmlc/xgboost/files/1320558/2017-09-21_xgboost_jvm_build_failure.txt

Must I add any further options to maven e.g. *-X*?

Thanks @superbobry!

**Comment by learn2Pro** [comment]
does anyone know the solution?

**Comment by frenzykryger** [comment]
I encountered this problem when running xgboost4j-spark compiled with OpenMP support  on two nodes with nthread=1 setting.
Excerpt from logs:
```
[10:34:24] Writing row.page to /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0803/container_e02_1518863324603_0803_01_000002/tmp/2-cache-02549934861360261688/train in 86.5872 MB/s, 4523 written
[10:34:25] Writing row.page to /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0803/container_e02_1518863324603_0803_01_000002/tmp/2-cache-13366236668570863199/train in 85.3678 MB/s, 4444 written
[10:34:27] SparsePage::Writer Finished writing to /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0803/container_e02_1518863324603_0803_01_000002/tmp/2-cache-02549934861360261688/train.row.page
[10:34:27] SparsePageSource: Finished writing to /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0803/container_e02_1518863324603_0803_01_000002/tmp/2-cache-02549934861360261688/train
[10:34:28] SparsePage::Writer Finished writing to /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0803/container_e02_1518863324603_0803_01_000002/tmp/2-cache-02549934861360261688/test.row.page
[10:34:28] SparsePageSource: Finished writing to /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0803/container_e02_1518863324603_0803_01_000002/tmp/2-cache-02549934861360261688/test
AssertError:maxdf must be smaller than FDSETSIZE


	at org.apache.spark.TaskFailedListener.onTaskEnd(SparkParallelismTracker.scala:116)
	at org.apache.spark.scheduler.SparkListenerBus$class.doPostEvent(SparkListenerBus.scala:45)
	at org.apache.spark.scheduler.LiveListenerBus.doPostEvent(LiveListenerBus.scala:36)
	at org.apache.spark.scheduler.LiveListenerBus.doPostEvent(LiveListenerBus.scala:36)
	at org.apache.spark.util.ListenerBus$class.postToAll(ListenerBus.scala:63)
	at org.apache.spark.scheduler.LiveListenerBus.postToAll(LiveListenerBus.scala:36)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(LiveListenerBus.scala:94)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(LiveListenerBus.scala:79)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(LiveListenerBus.scala:79)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1.apply$mcV$sp(LiveListenerBus.scala:78)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1279)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1.run(LiveListenerBus.scala:77)
```

assert comes from [rabit](https://github.com/dmlc/rabit/blob/a9a2a69dc1144180a43f7d2d1097264482be7817/src/socket.h#L511).

Environment:
OS:  Red Hat Enterprise Linux Server release 7.3 (Maipo)
xgboost built from 443ff746e9723dcf571769b0d6ea28fbcb3e4a3f
with following make/config.mk:
```
CC=gcc
CXX=g++
MPICXX=mpicxx

# the additional link flags you want to add
ADD_LDFLAGS =-L/usr/hdp/2.6.4.0-91/usr/lib

# the additional compile flags you want to add
ADD_CFLAGS = -I/usr/hdp/2.6.4.0-91/usr/include

# Whether enable openmp support, needed for multi-threading.
USE_OPENMP = 1

# whether use HDFS support during compile
USE_HDFS = 1

# whether use AWS S3 support during compile
USE_S3 = 0

# whether use Azure blob support during compile
USE_AZURE = 0
```

gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
g++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)

config in xgboost/jvm-packages/create_jni.py:
```
CONFIG = {
    "USE_OPENMP": "ON",
    "USE_HDFS": "ON",
    "USE_AZURE": "OFF",
    "USE_S3": "OFF",
    "HDFS_INCLUDE_DIR": "/usr/hdp/2.6.4.0-91/usr/include/",
    "HDFS_LIB": "/usr/hdp/2.6.4.0-91/usr/lib/libhdfs.so",
    "PLUGIN_UPDATER_GPU": "OFF",
    "JVM_BINDINGS": "ON"
}
```
OpenMP:
libgomp.x86_64           4.8.5-16.el7_4.2

**Comment by frenzykryger** [comment]
Stacktrace:
```
	at org.apache.spark.TaskFailedListener.onTaskEnd(SparkParallelismTracker.scala:116)
	at org.apache.spark.scheduler.SparkListenerBus$class.doPostEvent(SparkListenerBus.scala:45)
	at org.apache.spark.scheduler.LiveListenerBus.doPostEvent(LiveListenerBus.scala:36)
	at org.apache.spark.scheduler.LiveListenerBus.doPostEvent(LiveListenerBus.scala:36)
	at org.apache.spark.util.ListenerBus$class.postToAll(ListenerBus.scala:63)
	at org.apache.spark.scheduler.LiveListenerBus.postToAll(LiveListenerBus.scala:36)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(LiveListenerBus.scala:94)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(LiveListenerBus.scala:79)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(LiveListenerBus.scala:79)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1$$anonfun$run$1.apply$mcV$sp(LiveListenerBus.scala:78)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1279)
	at org.apache.spark.scheduler.LiveListenerBus$$anon$1.run(LiveListenerBus.scala:77)
```

**Comment by frenzykryger** [comment]
C stack trace:
```
Stack trace returned 10 entries:
[bt] (0) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN4dmlc15LogMessageFatalD1Ev+0x29) [0x7fa0d8165649]
[bt] (1) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit5utils17HandleAssertErrorEPKc+0x6e) [0x7fa0d81720fe]
[bt] (2) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit5utils6AssertEbPKcz+0xee) [0x7fa0d817222e]
[bt] (3) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit6engine13AllreduceBase16TryAllreduceTreeEPvmmPFvPKvS2_iRKN3MPI8DatatypeEE+0x6cd) [0x7fa0d832566d]
[bt] (4) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit6engine13AllreduceBase12TryAllreduceEPvmmPFvPKvS2_iRKN3MPI8DatatypeEE+0x12) [0x7fa0d8326882]
[bt] (5) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit6engine15AllreduceRobust11RecoverExecEPvmii+0x89) [0x7fa0d83340c9]
[bt] (6) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit6engine15AllreduceRobust8ShutdownEv+0x1c) [0x7fa0d83342fc]
[bt] (7) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(_ZN5rabit6engine8FinalizeEv+0x44) [0x7fa0d831d664]
[bt] (8) /storage/hadoop-host/yarn/local/usercache/tvsteam/appcache/application_1518863324603_0811/container_e02_1518863324603_0811_01_000002/tmp/libxgboost4j2277849048169570296.so(Java_ml_dmlc_xgboost4j_java_XGBoostJNI_RabitFinalize+0x9) [0x7fa0d81600f9]
[bt] (9) [0x7fa108be9a34]

terminate called after throwing an instance of 'dmlc::Error'
  what():  [03:22:02] /home/tvsteam/xgboost/xgboost/rabit/include/rabit/./internal/.././internal/utils.h:68: AssertError:maxdf must be smaller than FDSETSIZE
```

**Comment by frenzykryger** [comment]
@learn2Pro 
I've managed to work around this.
I discovered that this error just hides real error causing termination of process. 
I've launched spark job with nWorkers set to 1 and got different error, causing xgboost training process to crash - in my case this was issue #1869 (xgboost4j-spark crashes on Float.NaN in data). So I replaced NaNs (which I used to mark missing values) with another value and training finished successfully both when nWorkers was 1 and when there were >1 workers.

So this issue has following mechanic:
1. some error occurs during training process causing program to crash, but report about this error is swallowed and not shown in logs.
2. AllreduceBase::TryAllreduceTree tries to collect results, but previous error somehow breaks invariant of SelectHelper::Select call causing program to exit immediately with error above.

TL;DR

Run spark job with just one worker to see real root cause and fix it.
